pkg/reconciler/pipelinerun/resources/apply_test.go

```diff
@@ -2034,6 +2034,284 @@ func TestApplyParameters_ArrayIndexing(t *testing.T) {
+	ctx := context.Background()
+	cfg := config.FromContextOrDefaults(ctx)
+	cfg.FeatureFlags.EnableAPIFields = config.AlphaAPIFields
+	ctx = config.ToContext(ctx, cfg)
+	for _, tt := range []struct {
+		name     string
+		original v1beta1.PipelineSpec
+		params   []v1beta1.Param
+		expected v1beta1.PipelineSpec
+	}{{
+		name: "matrix params - array replacement",
+		original: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "first-param", Type: v1beta1.ParamTypeArray, Default: v1beta1.NewStructuredValues("default", "array", "value"),
+			}, {
+				Name: "second-param", Type: v1beta1.ParamTypeArray,
+			}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Params: []v1beta1.Param{{
+						Name: "first-matrix-task-first-param", Value: *v1beta1.NewStructuredValues("firstelement", "$(params.first-param)"),
+					}, {
+						Name: "first-matrix-task-second-param", Value: *v1beta1.NewStructuredValues("firstelement", "$(params.second-param[0])"),
+					}},
+				},
+			}},
+		},
+		params: []v1beta1.Param{
+			{Name: "second-param", Value: *v1beta1.NewStructuredValues("second-value", "array")},
+		},
+		expected: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{
+				{Name: "first-param", Type: v1beta1.ParamTypeArray, Default: v1beta1.NewStructuredValues("default", "array", "value")},
+				{Name: "second-param", Type: v1beta1.ParamTypeArray},
+			},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Params: []v1beta1.Param{
+						{Name: "first-matrix-task-first-param", Value: *v1beta1.NewStructuredValues("firstelement", "default", "array", "value")},
+						{Name: "first-matrix-task-second-param", Value: *v1beta1.NewStructuredValues("firstelement", "second-value")},
+					},
+				},
+			}},
+		},
+	}, {
+		name: "matrix include params - string replacement",
+		original: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "first-param", Type: v1beta1.ParamTypeString, Default: v1beta1.NewStructuredValues("default-value"),
+			}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Include: []v1beta1.MatrixInclude{{
+						Name: "build-1",
+						Params: v1beta1.Params{{
+							Name: "first-param", Value: *v1beta1.NewStructuredValues("$(params.first-param)"),
+						}}},
+					}},
+			}},
+		},
+		params: v1beta1.Params{{Name: "first-param", Value: *v1beta1.NewStructuredValues("default-value")}},
+		expected: v1beta1.PipelineSpec{
+			Params: []v1beta1.ParamSpec{{
+				Name: "first-param", Type: v1beta1.ParamTypeString, Default: v1beta1.NewStructuredValues("default-value"),
+			}},
+			Tasks: []v1beta1.PipelineTask{{
+				Matrix: &v1beta1.Matrix{
+					Include: []v1beta1.MatrixInclude{{
+						Name: "build-1",
+						Params: v1beta1.Params{{
+							Name: "first-param", Value: *v1beta1.NewStructuredValues("default-value"),
+						}}},
+					}},
+			}},
+		},
+	}, {
+		name: "matrix include params - string replacement - override default",
```

Comment:
what does `override default` mean in this case? Sorry I may get lost in the long cases 😄  Which default value is overide?